File: joss/comparisons/cat-5-schema-composition/20-implicit-type-coercion/example-implicit-type-coercion.py
"""
Error Class 20: Implicit Type Coercion in Row Binding

When concatenating DataFrames where the same column has different types,
pandas silently coerces to 'object' dtype. Subsequent numeric operations
fail silently or produce errors only at runtime.
"""
import pandas as pd

numeric_doses = pd.DataFrame({
    "drug": ["Aspirin", "Lisinopril"],
    "dose": [325, 10],
})

text_doses = pd.DataFrame({
    "drug": ["Insulin", "Warfarin"],
    "dose": ["sliding scale", "per INR"],
})

# SILENT: concat coerces 'dose' from int64 to object — no warning
combined = pd.concat([numeric_doses, text_doses], ignore_index=True)
print(f"dose dtype: {combined['dose'].dtype}")  # object

# SILENT: Arithmetic on mixed-type column
# combined["doubled"] = combined["dose"] * 2
# This would produce: 650, 20, "sliding scalesliding scale", "per INRper INR"
# String repetition instead of multiplication — silently wrong

File: joss/comparisons/cat-5-schema-composition/20-implicit-type-coercion/example-implicit-type-coercion.R
# Error Class 20: Implicit Type Coercion in Row Binding
#
# When binding rows from tibbles where the same column has different types,
# R silently coerces to the more general type (character). Subsequent
# numeric operations fail or produce NA.

library(tidyverse)

numeric_doses <- tibble(
  drug = c("Aspirin", "Lisinopril"),
  dose = c(325, 10)
)

text_doses <- tibble(
  drug = c("Insulin", "Warfarin"),
  dose = c("sliding scale", "per INR")
)

# SILENT: bind_rows coerces 'dose' to character — no error, no warning
combined <- bind_rows(numeric_doses, text_doses)
print(class(combined$dose))  # character
print(combined$dose)  # "325", "10", "sliding scale", "per INR"

# Arithmetic now fails or produces NA
# combined %>% mutate(doubled = as.numeric(dose) * 2)
# "sliding scale" → NA, "per INR" → NA (with warning)
